---
title: "sc gene expression quantification and cell type determination"
output: html_notebook
author: "Dominik Burri"
date: 2019-10-31
---

Setup
```{r setup, include=FALSE, cache = FALSE}
library("knitr")
## setting working directory
opts_knit$set(root.dir = "~/SCUREL")
```

# User settings
```{r}
cellranger_dir <- '.'
out_dir <- 'results/luad_matched/cell_type_annotation'
# out_dir <- 'activated_tcells/cell_type_annotation'
# out_dir <- 'results/spermatocytes/cell_type_annotation'
# either one dataset or two datasets (Lambrechts, Laughney, in this order)
# samples.loc <- 'datasets/spermatocytes/PRJNA413049/SraRunTable.cellranger.tsv'
samples.loc <- c('datasets/lambrechts/samples.luad_matched.tsv',
                 'datasets/laughney/samples.luad_matched.tsv')
marker.gene.loc <- 'resources/marker_genes.tsv'
```

## Install seurat

```{r}
# install.packages('Seurat')
library(Seurat)
library(dplyr)
library(ggplot2)
library(RColorBrewer)
library(biomaRt)

if (!requireNamespace("BiocManager", quietly = TRUE)){
  install.packages("BiocManager")
}
# if necessary, install packages below
#BiocManager::install("biomaRt")
#BiocManager::install('limma')
library(limma)
```

Overall plotting settings
```{r}
mytheme <- theme(plot.title = element_blank(),
        axis.text = element_text(size=8),
        axis.title = element_text(size=10),
        legend.title = element_text(size=10),
        legend.text = element_text(size=8))
```

# Introduction
This script implements the quantification and cell clustering method described in Lambrechts et al. 2018.

## canonical gene markers

```{r load_tables}
df <- read.table(marker.gene.loc, 
                 header = TRUE,
                 sep = "\t", stringsAsFactors = FALSE)

if(length(samples.loc) == 1){
  samples.tbl <- read.table(samples.loc,
                      header = TRUE, sep = '\t',
                      stringsAsFactors = FALSE)
}
if(length(samples.loc) == 2){
  lambrechts.tbl <- cbind(dataset = "Lambrechts", read.table(samples.loc[1],
                      header = TRUE, sep = '\t',
                      stringsAsFactors = FALSE))

laughney.tbl <- cbind(dataset = "Laughney", read.table(samples.loc[2],
                      header = TRUE, sep = '\t',
                      stringsAsFactors = FALSE))

samples.tbl <- full_join(lambrechts.tbl, laughney.tbl)
}
```


# Loading the data

```{r}
s_list <- samples.tbl$sample

named_list <- file.path(cellranger_dir, 
                        paste0("cellranger_count/", s_list, "/outs/filtered_feature_bc_matrix"))

names(named_list) <- s_list
```

Only when new data is used! Otherwise, load .RData object below.
Load previous run: approx. line 900
```{r eval=FALSE}
# Convert so seurat object
if(!file.exists(file.path(out_dir, 'seurat_object.rds'))){
  samples <- Read10X(data.dir = named_list)
  seurat_object <- CreateSeuratObject(counts = samples)
  saveRDS(seurat_object,
          file = file.path(out_dir, 'seurat_object.rds'))
} else{
  seurat_object <- readRDS(file.path(out_dir, 'seurat_object.rds'))
}

```

# Process cells

```{r}
# Percent mitochondrial for human genome
seurat_object[["percent.mt"]] <- PercentageFeatureSet(seurat_object, pattern = "^MT-")
# Percent mitochondrial for mouse genome
# seurat_object[["percent.mt"]] <- PercentageFeatureSet(seurat_object, pattern = "^mt-|^Mrp")
# seurat_object[["percent.ribo"]] <- PercentageFeatureSet(seurat_object, pattern = "^Rr")

d <- VlnPlot(seurat_object, 
        features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.ribo"), 
        ncol = 2,
        pt.size = 0,
        idents=s_list,
        log = FALSE)
ggsave(filename=file.path(out_dir, "violin_plots.png"), 
       plot=d,
       device = "png",
       width=20, height=20, units="cm",
       dpi=500)

# skip the subsetting if filtered counts are loaded (either redundant or some cells get removed)
seurat_object <- subset(seurat_object,
                        nCount_RNA > 200 &
                          (nFeature_RNA < 6000 & nFeature_RNA > 100) &
                          percent.mt < 10)

d <- VlnPlot(seurat_object,
        features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.ribo"),
        ncol = 2,
        pt.size = 0,
        idents=s_list,
        log = FALSE)
ggsave(filename=file.path(out_dir, "violin_plots_subsetted.png"),
       plot=d,
       device = "png",
       width=30, height=30, units="cm",
       dpi=500)
```

```{r}
VlnPlot(seurat_object,
        features = 'nCount_RNA',
        group.by = 'orig.ident',
        pt.size = 0,
        log = TRUE) +
  theme(legend.position = 'none')
ggsave(file.path(out_dir, 'vlnplot_counts_per_sample.png'),
       width = 15, height = 8, units = 'cm')

VlnPlot(seurat_object,
        features = 'nFeature_RNA',
        group.by = 'orig.ident',
        pt.size = 0,
        log = TRUE) +
  theme(legend.position = 'none')
ggsave(file.path(out_dir, 'vlnplot_genes_per_sample.png'),
       width = 15, height = 8, units = 'cm')
```

# Feature selection

```{r}
#remove genes with zero variance
seurat_object <- seurat_object[ Matrix::rowSums(seurat_object) > 0, ]
```

Save gene list to file
```{r}
write.table(rownames(seurat_object),
            file = file.path(out_dir, "genes_considered.tsv"),
            quote = FALSE, sep = "\t",
            row.names = FALSE, col.names = FALSE)
```


## Normalise 

```{r}
seurat_object <- NormalizeData(seurat_object)
```

## Select variably expressed genes

```{r}
seurat_object <- FindVariableFeatures(seurat_object, selection.method = "vst",
                                               nfeatures = 2192,
                                               mean.cutoff = c(0.125, 3),
                                      dispersion.cutoff = c(0.5, Inf))
top20 <- head(VariableFeatures(seurat_object), 20)
d <- LabelPoints(plot = VariableFeaturePlot(seurat_object),
            points = top20, repel = TRUE)
ggsave(filename=file.path(out_dir, "variable_features.png"),
       plot=d,
       device = "png",
       width=18, height=10, units="cm",
       dpi=500)
```

Scale features
```{r}
seurat_object <- ScaleData(seurat_object,
                           features = VariableFeatures(object = seurat_object),
                           vars.to.regress = NULL,
                           model.use = 'poisson',
                           do.center = TRUE,
                           do.scale = TRUE)
```

## Add metadata
* patient
* tissue orgin
```{r}
get_patient <- function(x){
        return(samples.tbl[samples.tbl$sample == x, 'patient'])
}

seurat_object@meta.data$patient <- as.factor(sapply(seurat_object@meta.data$orig.ident, function(x) get_patient(x)))

find_origin <- function(orig.ident){
  return(samples.tbl[samples.tbl$sample == as.character(orig.ident),'origin'])
}

seurat_object[["origin"]] <- apply(seurat_object[["orig.ident"]], 1, function(x) find_origin(x))

get_dataset <- function(orig.ident){
  return(samples.tbl[samples.tbl$sample == as.character(orig.ident), 'dataset'])
}
# add dataset info into seurat object
if(length(samples.loc) == 2){
  seurat_object[["dataset"]] <- apply(seurat_object[["orig.ident"]], 1, function(x) get_dataset(x))
}

```

How many cells?
```{r}
table(seurat_object[["dataset"]])
table(seurat_object[["origin"]])
table(seurat_object[["orig.ident"]])
```


# Perform linear dimensional reduction
```{r}
seurat_object <- RunPCA(seurat_object, npcs = 30,
                        features = VariableFeatures(object = seurat_object))
```

Plot PCA dimensions
```{r}
DimPlot(seurat_object, reduction = 'pca', dims = c(1,2),
        group.by = "orig.ident")
ggsave(file.path(out_dir, "pca_12_samples.png"),
       width = 20, height = 10, units = 'cm')
DimPlot(seurat_object, reduction = 'pca', dims = c(3,4),
        group.by = "orig.ident")
ggsave(file.path(out_dir, "pca_34_samples.png"),
       width = 20, height = 10, units = 'cm')

DimPlot(seurat_object, reduction = 'pca', dims = c(1,2),
        group.by = "dataset") +
  mytheme +
  labs(x="PC 1", y = "PC 2") + 
  theme(legend.position = "none")
ggsave(file.path(out_dir, "pca_12_dataset.png"),
       width = 7, height = 7, units = 'cm')
DimPlot(seurat_object, reduction = 'pca', dims = c(3,4),
        group.by = "dataset")+
  mytheme +
  labs(x="PC 1", y = "PC 2") + 
  theme(legend.position = "none")
ggsave(file.path(out_dir, "pca_34_dataset.png"),
       width = 7, height = 7, units = 'cm')
```

Decide on nr. of dimensions
```{r}
ElbowPlot(seurat_object)
ggsave(file.path(out_dir, "elbow_plot.png"),
       width = 10, height = 10, units = 'cm')

VizDimLoadings(seurat_object, dims = 1:10, reduction = "pca")

DimHeatmap(seurat_object, dims = 1:20, cells = 500, balanced = TRUE)
```

Choose the number of dimensions: 
```{r}
n_dims <- 8
```

LUAD: 10
Activated T cells: 5
Spermatocytes: 8


## Plot tSNE
Summarise the first n_dims PC further by tSNE dimensionality reduction.
```{r}
seurat_object <- RunTSNE(seurat_object,
                         dims = 1:n_dims,
                         reduction = "pca",
                         check_duplicates=FALSE,
                         seed.use = 1)
seurat_object <- RunUMAP(seurat_object, 
                         dims = 1:n_dims,
                        n.components = 2,
                        n.neighbors = 20,
                        spread = 1,
                        repulsion.strength = 1,
                        min.dist= .1,
                        verbose = F,
                        n.epochs = 200,
                        metric = "euclidean",
                        seed.use = 42)

```

### plot tSNE per sample
```{r}
# plot tSNE with different groupings
q <- DimPlot(seurat_object, reduction = "tsne", group.by = "orig.ident") + mytheme
ggsave(filename=file.path(out_dir, "tsne_sample.png"), 
       plot=q,
       device = "png",
       width=12, height=11, units="cm")
q <- DimPlot(seurat_object, reduction = "umap", group.by = "orig.ident") + mytheme
ggsave(filename=file.path(out_dir, "umap_sample.png"), 
       plot=q,
       device = "png",
       width=12, height=11, units="cm")

q <- DimPlot(seurat_object, reduction = "tsne", group.by = "dataset") + mytheme
ggsave(filename=file.path(out_dir, "tsne_dataset.png"), 
       plot=q,
       device = "png",
       width=12, height=11, units="cm")
q <- DimPlot(seurat_object, reduction = "umap", group.by = "dataset") + mytheme
ggsave(filename=file.path(out_dir, "umap_dataset.png"), 
       plot=q,
       device = "png",
       width=12, height=11, units="cm")

q <- DimPlot(seurat_object, reduction = "tsne", group.by = "origin") + mytheme
ggsave(filename=file.path(out_dir, "tsne_origin.png"), 
       plot=q,
       device = "png",
       width=12, height=11, units="cm")
q <- DimPlot(seurat_object, reduction = "umap", group.by = "origin") + mytheme
ggsave(filename=file.path(out_dir, "umap_origin.png"), 
       plot=q,
       device = "png",
       width=12, height=11, units="cm")

q <- DimPlot(seurat_object, reduction = "tsne", group.by = "patient") + mytheme
ggsave(filename=file.path(out_dir, "tsne_patients.png"), 
       plot=q,
       device = "png",
       width=12, height=11, units="cm")
q <- DimPlot(seurat_object, reduction = "umap", group.by = "patient") + mytheme
ggsave(filename=file.path(out_dir, "umap_patients.png"), 
       plot=q,
       device = "png",
       width=12, height=11, units="cm")
```

Do QC parameters or datasets influence separation of the cells?
```{r}
q <- FeaturePlot(seurat_object, 
                 features = c('percent.mt', 'percent.ribo'),
                 reduction = "umap",
                 order = TRUE)
ggsave(filename=file.path(out_dir, "umap_percent_mito_ribo.png"), 
       plot=q,
       device = "png",
       width=20, height=10, units="cm",
       dpi=500)
```

# Perform Dataset integration
If there are any batch effects in the data (i.e. no overlap of samples in UMAP in individual clusters), then dataset integration is necessary
```{r}
#library(devtools)
#install_github("immunogenomics/harmony")
library(harmony)

seurat_object <- RunHarmony(
  seurat_object,
  reduction = "pca", dims.use = 1:30,
  nclust = NULL,
  group.by.vars = "dataset",
  project.dim = TRUE)
```

```{r}
DimPlot(seurat_object, reduction = "harmony", dims = c(1,2))
ggsave(file.path(out_dir, "harmony_12_samples.png"),
       width = 20, height = 10, units = 'cm')
DimPlot(seurat_object, reduction = "harmony", dims = c(3,4))
ggsave(file.path(out_dir, "harmony_34_samples.png"),
       width = 20, height = 10, units = 'cm')

DimPlot(seurat_object, reduction = "harmony", 
        dims = c(1,2), 
        group.by = "dataset") +
  mytheme +
  labs(x="Harmony 1", y = "Harmony 2") + 
  theme(legend.position = "none")
ggsave(file.path(out_dir, "harmony_12_dataset.png"),
       width = 7, height = 7, units = 'cm')

DimPlot(seurat_object, reduction = "harmony", 
        dims = c(3,4), 
        group.by = "dataset")+
  mytheme +
  labs(x="Harmony 3", y = "Harmony 4") + 
  theme(legend.position = "none")
ggsave(file.path(out_dir, "harmony_34_dataset.png"),
       width = 7, height = 7, units = 'cm')

ElbowPlot(seurat_object, ndims = 20, reduction = "harmony")
ggsave(file.path(out_dir, "elbow_plot_harmony.png"),
       width = 10, height = 10, units = 'cm')
```

```{r}
seurat_object <- RunUMAP(object = seurat_object,
                        reduction = "harmony",
                        dims = 1:n_dims,
                        n.components = 2,
                        n.neighbors = 20,
                        spread = 1,
                        repulsion.strength = 1,
                        min.dist= .1,
                        verbose = F,
                        n.epochs = 200,
                        metric = "euclidean",
                        seed.use = 42,
                        reduction.name="umap.harmony",
                        reduction.key = "umapharmony_")
seurat_object <- RunTSNE(object = seurat_object,
                        reduction = "harmony",
                        perplexity=30,
                        max_iter=1000,
                        theta=0.5,
                        eta=200,
                        exaggeration_factor=12,
                        dims.use = 1:n_dims,
                        verbose = F,
                        num_threads=0,
                        seed.use = 42,
                        reduction.name = "tsne.harmony",
                        reduction.key = "tsneharmony_")
```

```{r}
q <- DimPlot(seurat_object, reduction = "tsne.harmony", group.by = "orig.ident") + mytheme
ggsave(filename=file.path(out_dir, "tsne_harmony_sample.png"), 
       plot=q,
       device = "png",
       width=12, height=11, units="cm")
q <- DimPlot(seurat_object, reduction = "umap.harmony", group.by = "orig.ident") + mytheme
ggsave(filename=file.path(out_dir, "umap_harmony_sample.png"), 
       plot=q,
       device = "png",
       width=12, height=11, units="cm")

q <- DimPlot(seurat_object, reduction = "tsne.harmony", group.by = "dataset") + mytheme
ggsave(filename=file.path(out_dir, "tsne_harmony_dataset.png"), 
       plot=q,
       device = "png",
       width=12, height=11, units="cm")
q <- DimPlot(seurat_object, reduction = "umap.harmony", group.by = "dataset") + mytheme
ggsave(filename=file.path(out_dir, "umap_harmony_dataset.png"), 
       plot=q,
       device = "png",
       width=12, height=11, units="cm")

q <- DimPlot(seurat_object, reduction = "tsne.harmony", group.by = "patient") + mytheme
ggsave(filename=file.path(out_dir, "tsne_harmony_patients.png"), 
       plot=q,
       device = "png",
       width=12, height=11, units="cm")
q <- DimPlot(seurat_object, reduction = "umap.harmony", group.by = "patient") + mytheme
ggsave(filename=file.path(out_dir, "umap_harmony_patients.png"), 
       plot=q,
       device = "png",
       width=12, height=11, units="cm")

rn <- DimPlot(seurat_object_renamed, reduction = "umap.harmony",
        group.by = "origin") + mytheme
ggsave(filename=file.path(out_dir, "umap_harmony_origin.png"), 
       plot=rn,
       device = "png",
       width=12, height=11, units="cm")
```

Do QC parameters or datasets influence separation of the cells?
```{r}
q <- FeaturePlot(seurat_object, features = c('percent.mt', 'percent.ribo'),
                 reduction = "umap.harmony") + mytheme
ggsave(filename=file.path(out_dir, "umap_harmony_percent_mito_ribo.png"), 
       plot=q,
       device = "png",
       width=20, height=10, units="cm",
       dpi=500)
```
## Save seurat object to file
```{r}
if(!file.exists(file.path(out_dir, "seurat_object_processed.rds"))){
  saveRDS(seurat_object, file=file.path(out_dir, "seurat_object_processed.rds"),
     compress = TRUE)
} else{
  seurat_object <- readRDS(file.path(out_dir, "seurat_object_processed.rds"))  
}
```

# Cluster the cells
Based on the description given by Bernhard in email from 5.11.2019
```{r}
seurat_object <- FindNeighbors(seurat_object, reduction = "pca", dims = 1:n_dims,
                               k.param = 15)
seurat_object <- FindClusters(seurat_object, resolution = 0.3)
```


Plot clustering result in 2D
```{r}
d <- DimPlot(seurat_object, reduction = "tsne", label = TRUE, group.by = "seurat_clusters") + mytheme + theme(legend.position = "none")
ggsave(filename=file.path(out_dir, "tsne_clusters.png"), 
       plot=d,
       device = "png",
       width=10, height=11, units="cm")
d <- DimPlot(seurat_object, reduction = "umap", label = TRUE, group.by = "seurat_clusters") + mytheme + theme(legend.position = "none")
ggsave(filename=file.path(out_dir, "umap_clusters.png"), 
       plot=d,
       device = "png",
       width=12, height=11, units="cm")

d <- DimPlot(seurat_object, reduction = "tsne.harmony", label = TRUE, group.by = "seurat_clusters") + mytheme + theme(legend.position = "none")
ggsave(filename=file.path(out_dir, "tsne_harmony_clusters.png"), 
       plot=d,
       device = "png",
       width=12, height=11, units="cm")
d <- DimPlot(seurat_object, reduction = "umap.harmony", label = TRUE, 
             group.by = "seurat_clusters", label.size = 3) + 
  mytheme + 
  theme(legend.position = "none") + 
  labs(x = "UMAP 1", y="UMAP 2") + 
  theme(axis.text = element_blank(),
        axis.ticks = element_blank())
ggsave(filename=file.path(out_dir, "umap_harmony_clusters.png"), 
       plot=d,
       device = "png",
       width=6, height=7, units="cm")
```

```{r}
d <- DimPlot(seurat_object, 
        reduction = "umap.harmony", 
        group.by = "seurat_clusters",
        split.by = 'dataset',
        label = TRUE) + mytheme + theme(legend.position = "none")
ggsave(filename=file.path(out_dir, "umap_harmony_clusters_per_dataset.png"), 
       plot=d,
       device = "png",
       width=18, height=11, units="cm",
       dpi=500)
```

## plot expression of marker genes
```{r}
if(!dir.exists(file.path(out_dir, "marker_expression"))){
  dir.create(file.path(out_dir, "marker_expression"))
}
for(ft.to.plot in df$marker_gene){
  if(!ft.to.plot %in% rownames(seurat_object)){
    next
  }
  p <- FeaturePlot(seurat_object, features = ft.to.plot,
                 reduction = "umap",
                 order = TRUE)

  ggsave(filename=file.path(out_dir, "marker_expression", paste0(ft.to.plot, ".png")), 
       plot=p,
       device = "png",
       width=10, height=10, units="cm",
       dpi=300)
}
```

```{r}
ticks <- levels(seurat_object$seurat_clusters)
names(ticks) <- ticks
ticks[seq(2, length(ticks), by = 2)] <- ""
d <- DotPlot(object = seurat_object,
        features = df$marker_gene,
        group.by = "seurat_clusters",
        dot.scale = 2) + 
  coord_flip() + 
  theme(axis.text = element_text(size=8),
        axis.text.x = element_text(angle = 90),
        axis.title = element_blank(),
        legend.text = element_text(size=8),
        legend.title = element_blank())
ggsave(filename=file.path(out_dir, "dot_plot_marker_genes.png"), 
       plot=d,
       device = "png",
       width=11, height=8, units="cm")
ggsave(filename=file.path(out_dir, "dot_plot_marker_genes.pdf"), 
       plot=d,
       device = "pdf",
       width=11, height=8, units="cm")

 d <- DoHeatmap(object = seurat_object,
          features = df$marker_gene,
          group.by = "seurat_clusters")
ggsave(filename=file.path(out_dir, "heatmap_marker_genes.png"), 
       plot=d,
       device = "png",
       width=30, height=20, units="cm",
       dpi=300)
```

## plot marker expression per cluster
```{r}
if(!dir.exists(file.path(out_dir, "marker_expression_per_cluster"))){
  dir.create(file.path(out_dir, "marker_expression_per_cluster"))
}
for(ft.to.plot in df$marker_gene){
  if(!ft.to.plot %in% rownames(seurat_object)){
    next
  }
  v <- VlnPlot(seurat_object, features = ft.to.plot, 
             pt.size = 0.1) + 
    theme(legend.position = 'none')
  ggsave(filename=file.path(out_dir, "marker_expression_per_cluster", paste0(ft.to.plot, ".png")), 
       plot=v,
       device = "png",
       width=12, height=10, units="cm",
       dpi=300)
}

```

# Find differentially expressed genes

```{r}
max_cells <- max(100, min(table(seurat_object$seurat_clusters)))
seurat_object.markers <- FindAllMarkers(seurat_object, 
                                          features = NULL,
                                          min.pct = 0.1, 
                                          logfc.threshold = 0.25,
                                          only.pos = TRUE,
                                          max.cells.per.ident = max_cells,
                                          min.diff.pct = -Inf)
```

```{r}
getwd()
if(!file.exists(file.path(out_dir, "seurat_object_markers.rds"))){
  saveRDS(seurat_object.markers, 
          file=file.path(out_dir, "seurat_object_markers.rds"))
} else{
  seurat_object.markers <- readRDS(file.path(out_dir, "seurat_object_markers.rds"))
}
```

```{r}
top5 <- seurat_object.markers %>% group_by(cluster) %>% filter(p_val_adj < 0.01) %>% top_n(n=5, wt=avg_logFC)
top5
```

```{r}
d <- DotPlot(object = seurat_object,
        features = as.character(unique(top5$gene)),
        group.by = "seurat_clusters") + coord_flip()
ggsave(filename=file.path(out_dir, "dot_plot_top5.png"), 
       plot=d,
       device = "png",
       width=22, height=30, units="cm",
       dpi=300)
```

## Assign cell type identiy to clusters
First, check the table for the given marker genes
```{r}
somf <- seurat_object.markers %>% filter(gene %in% df$marker_gene, p_val_adj < 0.01)

# based on the gene, set cell type
somf$cell_type <- sapply(somf$gene, function(x) df[df["marker_gene"]==x,"cell_type"])
# somf %>% filter(cluster == 14)
somf
```

Together with dot_plot_marker_genes.png and marker_expression_per_cluster.png plot, decide on cell type for each cluster.
If cell type cannot be assigned to a cluster, set "unassigned", we will come back to these clusters.

# automate cluster assignment
Criteria for cell type assignment:
1. >= 3/4 marker genes are expressed in cluster => fraction >= 0.6 (# cell type markers) / # markers in this cluster
2. for cancer only EPCAM is expressed and nothing else
3. if there are multiple marker genes expressed (i.e. from different cell types), the one satisfying 1. is chosen. OR: fraction >= 0.6
```{r}
tmp <- somf %>% group_by(cluster, cell_type) %>% summarise(n = n())
n_markers <- tmp %>% group_by(cluster) %>% summarise(n_markers = sum(n))
tmp <- full_join(tmp, n_markers, by = "cluster")
table <- mutate(tmp, fraction = n / n_markers)
table
# function for identifying cell type
get_cell_type <- function(table, clust){
  tmp <- filter(table, cluster == clust)
  ct <- "unassigned"
  if(nrow(tmp) != 0){
    if(any(tmp$fraction > 0.6)){
      ct <- filter(tmp, fraction > 0.6) %>% pull(cell_type)
    }
  }
  return(ct)
}
```

### obtain cell type annotation

```{r}
new.cluster.ids <- sapply(as.integer(levels(table$cluster)), function(x) get_cell_type(table, x))
table(new.cluster.ids)
```


There are some clusters left out, check the expression of their top marker genes.
At least 3 markers needed for assigning cell to particular cell type, else leave unassigned
```{r}
clusters_left <- seurat_object.markers %>% group_by(cluster) %>% filter(cluster %in% (which(new.cluster.ids == "unassigned")-1), p_val_adj < 0.01) %>% top_n(n=20, wt=avg_logFC)
clusters_left
```

## re-define clusters as cell types
based on the previous findings (MANUAL!)
And on Supplementary table 3: contains for each subcluster the list of found marker genes.
Criteria for cell type definition:
* adjusted p-value < 0.01
* at least 3 genes for a cell type are described in the suppl. table 3

if criteria not met, cluster remains unassigned 

### results for Lambrechts et al. 2018
```{r}
which(new.cluster.ids == "unassigned") - 1 # find cluster number 
new.cluster.ids[which(new.cluster.ids == "unassigned")] <- c("unassigned", "cancer", 
                                                             "cancer", "erythroblasts", "myeloid")
```

### results for Laughney et al. 2020
```{r}
which(new.cluster.ids == "unassigned") - 1 # find cluster number 
new.cluster.ids[which(new.cluster.ids == "unassigned")] <- c("unassigned", "T_cell", "unassigned",
                                                             "myeloid", "myeloid",
                                                             "alveolar")
```

### results for Lambrechts + Laughney LUAD
```{r}
which(new.cluster.ids == "unassigned") - 1 # find cluster number 
new.cluster.ids[which(new.cluster.ids == "unassigned")] <- c("T_cell", "T_cell", "cancer","cancer")
```

### results for activated T cells
Based on top5 marker genes per cluster, individually checked on mouse genome database http://www.informatics.jax.org and expression atlas https://www.ebi.ac.uk/gxa/home.
```{r}
new.cluster.ids <- c("naive_tcell", # Ccr7, Sell
                     "unassigned", 
                     "activated_tcell", # Pclaf, Hmgb2, Birc5, Ccl5, Uba52
                     "activated_tcell", # Ccl5, Uba52
                     "naive_tcell", # Ccr7, Sell
                     "activated_tcell") # Pclaf, Hmgb2, Birc5
table(seurat_object$seurat_clusters)
```

### results for spermatocytes
```{r}
new.cluster.ids <- c("unassigned", "unassigned", "unassigned",
                     "unassigned", "unassigned", "unassigned",
                     "unassigned", "condensing_spermatids", "unassigned",
                     "round_spermatids", "elongating_spermatids", "spermatocytes",
                     "spermatocytes", "unassigned")
```

rename cluster ids
```{r}
# remove clusters not assigned to any of the desired cell types > also possible later on
names(new.cluster.ids) <- levels(seurat_object)
seurat_object_renamed <- RenameIdents(seurat_object, new.cluster.ids)
# add metadata info
seurat_object_renamed[["cell_type"]] <- as.character(seurat_object_renamed@active.ident)

```

How many cells per cluster?
```{r}
table(seurat_object_renamed@meta.data$cell_type)
table(seurat_object_renamed@meta.data$cell_type, seurat_object_renamed@meta.data$orig.ident)
```

```{r save_renamed}
if(!file.exists(file.path(out_dir, "seurat_object_renamed.rds"))){
  saveRDS(seurat_object_renamed,
     file = file.path(out_dir, "seurat_object_renamed.rds"),
     compress = TRUE)
} else{
  seurat_object_renamed <- readRDS(file = file.path(out_dir, "seurat_object_renamed.rds"))
  # add all metainformation to previous seurat object
  seurat_object <- seurat_object_renamed
}
```

## Plot and save tsne & umap representation
```{r}
rn <- DimPlot(seurat_object_renamed, reduction = "tsne", label = TRUE,
        cols=rep(brewer.pal(9, "Set1"), times=2)) + mytheme + theme(legend.position = "none")
ggsave(filename=file.path(out_dir, "tsne_cell_types.png"), 
       plot=rn,
       device = "png",
       width=12, height=11, units="cm")
# UMAP
rn <- DimPlot(seurat_object_renamed, reduction = "umap", label = TRUE,
        cols=brewer.pal(9, "Set1")) + mytheme + theme(legend.position = "none")

ggsave(filename=file.path(out_dir, "umap_cell_types.png"), 
       plot=rn,
       device = "png",
       width=12, height=11, units="cm")

rn <- DimPlot(seurat_object_renamed, reduction = "tsne.harmony", label = TRUE,
        cols=rep(brewer.pal(9, "Set1"), times=2)) + mytheme
ggsave(filename=file.path(out_dir, "tsne_harmony_cell_types.png"), 
       plot=rn,
       device = "png",
       width=12, height=11, units="cm")
# UMAP
rn <- DimPlot(seurat_object_renamed, reduction = "umap.harmony", label = TRUE,
        cols=brewer.pal(9, "Set1")) + mytheme

ggsave(filename=file.path(out_dir, "umap_harmony_cell_types.png"), 
       plot=rn,
       device = "png",
       width=12, height=11, units="cm")
```

```{r}
ct_cols <- brewer.pal(9, "Set1")
# replace yellow
ct_cols[6] <- "#DCE621"
rn <- DimPlot(seurat_object_renamed, 
        reduction = "umap.harmony", 
        label = FALSE,
        cols=ct_cols,
        split.by = "dataset") + mytheme +
  labs(x = "UMAP 1", y="UMAP 2") + 
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.text = element_text(size=8),
        legend.position = "right")
ggsave(filename=file.path(out_dir, "umap_harmony_cell_types_per_dataset.png"), 
       plot=rn,
       device = "png",
       width=12, height=7, units="cm")
```

plot by different splits
```{r}
rn <- DimPlot(seurat_object_renamed, reduction = "umap.harmony", split.by = "patient", label = FALSE,
        cols=rep(brewer.pal(9, "Set1"), times=2),
        ncol = 2)
ggsave(filename=file.path(out_dir, "umap_harmony_per_patient.png"), 
       plot=rn,
       device = "png",
       width=30, height=45, units="cm",
       dpi=500)

rn <- DimPlot(seurat_object_renamed, 
              reduction = "umap.harmony", 
              split.by = "origin", 
              label.size = 6,
              cols=rep(brewer.pal(9, "Set1"), times=2))
ggsave(filename=file.path(out_dir, "umap_harmony_per_origin.png"), 
       plot=rn,
       device = "png",
       width=30, height=15, units="cm",
       dpi=500)

if(length(samples.loc) == 2){
  rn <- DimPlot(seurat_object_renamed, reduction = "umap", split.by = "dataset",
                label = TRUE, cols=rep(brewer.pal(9, "Set1"), times=2))
  ggsave(filename=paste0(out_dir, "umap_per_dataset.png"), 
       plot=rn,
       device = "png",
       width=30, height=15, units="cm",
       dpi=500)
}
rn
```

## Re-color cell types

### Spermatocytes
```{r}

cts <- c("unassigned", "unassigned", "unassigned",
                "elongating_spermatids", "spermatocytes")
names(cts) <- levels(seurat_object_renamed)
seurat_object_recolor <- RenameIdents(seurat_object_renamed, cts)
sperm_cols <- c(unassigned = "grey", spermatocytes = "orange", 
                elongating_spermatids = "purple")
rn <- DimPlot(seurat_object_recolor, reduction = "tsne", label = TRUE,
        cols=sperm_cols) + mytheme + theme(legend.position = "none")
ggsave(filename=file.path(out_dir, "tsne_cell_types_recolored.png"), 
       plot=rn,
       device = "png",
       width=12, height=11, units="cm")
# UMAP
rn <- DimPlot(seurat_object_recolor, reduction = "umap", 
              label = FALSE,
              pt.size = 0.1,
              cols=sperm_cols) + mytheme + 
  theme(legend.position = "none",
        axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.text = element_text(size=6))

ggsave(filename=file.path(out_dir, "umap_cell_types_recolored.png"), 
       plot=rn,
       device = "png",
       width=5, height=4.5, units="cm")
```

### Activated T cells
```{r}
tcell_cols <- brewer.pal(3, "Set1")
names(tcell_cols) <- c("activated_tcell", "unassigned", "naive_tcell")
tcell_cols[2] <- "grey"
rn <- DimPlot(seurat_object_renamed, reduction = "tsne.harmony", label = TRUE,
        cols=tcell_cols) + mytheme + theme(legend.position = "none")
ggsave(filename=file.path(out_dir, "tsne_harmony_cell_types_recolored.png"), 
       plot=rn,
       device = "png",
       width=12, height=11, units="cm")
# UMAP
rn <- DimPlot(seurat_object_renamed, reduction = "umap.harmony", 
              label = FALSE,
              pt.size = 0.1,
              cols=tcell_cols) + mytheme + 
  theme(legend.position = "none",
        axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.text = element_text(size=6))

ggsave(filename=file.path(out_dir, "umap_harmony_cell_types_recolored.png"), 
       plot=rn,
       device = "png",
       width=5, height=4.5, units="cm")
```

# Reads per cell
What is the median number of reads in cells of different types (cell size)?

Plot UMI count and gene count distribution per cell type
```{r}
meta.df <- seurat_object_renamed@meta.data
meta.df$id <- rownames(meta.df)
# seurat_d1 <- subset(seurat_object_renamed, cells = filter(meta.df) %>% pull(id))
VlnPlot(seurat_object_renamed,
        features = c('nFeature_RNA'),
        group.by = 'cell_type',
        split.by = 'dataset',
        pt.size = 0,
        log = FALSE)
ggsave(file.path(out_dir, 'vlnplot_genes_per_cell_type.png'),
       width = 20, height = 10, units = 'cm')

VlnPlot(seurat_object_renamed,
        features = c('nFeature_RNA'),
        group.by = 'cell_type',
        split.by = 'origin',
        pt.size = 0,
        log = FALSE)
ggsave(file.path(out_dir, 'vlnplot_genes_per_cell_type_origin.png'),
       width = 20, height = 10, units = 'cm')

VlnPlot(seurat_object_renamed,
        features = c('nCount_RNA'),
        group.by = 'cell_type',
        split.by = 'dataset',
        pt.size = 0,
        log = TRUE) + 
  mytheme +
  theme(axis.title = element_blank(),
        legend.position = "none")
ggsave(file.path(out_dir, 'vlnplot_counts_per_cell_type.png'),
       width = 14, height = 8, units = 'cm')

VlnPlot(seurat_object_renamed,
        features = c('nCount_RNA'),
        group.by = 'cell_type',
        split.by = 'origin',
        pt.size = 0,
        log = TRUE)
ggsave(file.path(out_dir, 'vlnplot_counts_per_cell_type_origin.png'),
       width = 20, height = 10, units = 'cm')
```

## plot marker expression per cell type
```{r}
ft.to.plot <- "CD47"
df.md <- seurat_object_renamed@meta.data
cells.alveolar.LB <- df.md %>% filter(cell_type == "alveolar" & origin == "normal" & dataset == "Lambrechts")
cells.alveolar.LN <- df.md %>% filter(cell_type == "alveolar" & origin == "normal" & dataset == "Laughney")
cells.cancer.LB <- df.md %>% filter(cell_type == "cancer" & origin == "tumor"  & dataset == "Lambrechts")
cells.cancer.LN <- df.md %>% filter(cell_type == "cancer" & origin == "tumor"  & dataset == "Laughney")
if(!dir.exists(file.path(out_dir, "marker_expression_per_cell_type"))){
  dir.create(file.path(out_dir, "marker_expression_per_cell_type"))
}


so.cells <- subset(seurat_object_renamed, cells = c(rownames(cells.alveolar.LB),
                                                    rownames(cells.cancer.LB)))
# differential expression (no downsampling)
dge <- FindMarkers(so.cells, ident.1 = "alveolar", ident.2 = "cancer",
            features = ft.to.plot,
            logfc.threshold = 0,
            test.use = "wilcox",
            min.pct = 0)

v <- VlnPlot(so.cells, 
             features = ft.to.plot, 
             pt.size = 0.1) +
  labs(caption = paste0("pct.cancer = ", dge$pct.2[1],
                        ", pct.alveolar = ", dge$pct.1[1],
                        "\navg_log2FC = ", signif(dge$avg_log2FC[1],3), 
                        ", p_val_adj = ", signif(dge$p_val_adj[1], 3))) +
  theme(legend.position = "none",
        axis.title.x = element_blank())
ggsave(filename=file.path(out_dir, "marker_expression_per_cell_type", 
                          paste0(ft.to.plot, "_Lambrechts.png")), 
     plot=v,
     device = "png",
     width=9, height=8, units="cm",
     dpi=300)

so.cells <- subset(seurat_object_renamed, cells = c(rownames(cells.alveolar.LN),
                                                    rownames(cells.cancer.LN)))
# differential expression (no downsampling)
dge <- FindMarkers(so.cells, ident.1 = "alveolar", ident.2 = "cancer",
            features = ft.to.plot,
            logfc.threshold = 0,
            test.use = "wilcox",
            min.pct = 0)

v <- VlnPlot(so.cells, 
             features = ft.to.plot, 
             pt.size = 0.1) +
  labs(caption = paste0("pct.cancer = ", dge$pct.2[1],
                        ", pct.alveolar = ", dge$pct.1[1],
                        "\navg_log2FC = ", signif(dge$avg_log2FC[1],3), 
                        ", p_val_adj = ", signif(dge$p_val_adj[1], 3))) +
  theme(legend.position = "none",
        axis.title.x = element_blank())
ggsave(filename=file.path(out_dir, "marker_expression_per_cell_type", 
                          paste0(ft.to.plot, "_Laughney.png")), 
     plot=v,
     device = "png",
     width=9, height=8, units="cm",
     dpi=300)


```

## PCA plots on cell types

```{r pca}
DimPlot(seurat_object_renamed, 
        reduction = 'pca',
        dims = c(1,2))
ggsave(file.path(out_dir, "pca_12_cell_types.png"),
       width = 15, height = 10, units = 'cm')
DimPlot(seurat_object_renamed, 
        reduction = 'pca', 
        dims = c(3,4))
ggsave(file.path(out_dir, "pca_34_cell_types.png"),
       width = 15, height = 10, units = 'cm')
```

# write cell barcode cluster identity to file
only include the cell barcodes that are identified as cells.
```{r}
cell_types <- seurat_object_renamed@meta.data %>% select(orig.ident, nCount_RNA, nFeature_RNA)
cell_types <- tibble::rownames_to_column(cell_types)
cell_types$cell <- sapply(cell_types$rowname, function(x) strsplit(x, "_")[[1]][2])
cell_types$cell_type <- as.character(Idents(seurat_object_renamed))
cell_types <- cell_types %>% select(orig.ident, cell, cell_type, nCount_RNA, nFeature_RNA)
write.table(cell_types, file=file.path(out_dir, "cell_type_annotations.csv"),
            quote = FALSE,
            sep = ",",
            row.names = FALSE)
```
##################################################
# plot bar chart

```{r}
# CELL_TYPES <- c("T cell", "B cell", "mast cell", "myeloid",                                                       "alveolar", "epithelial", "endothelial", "fibroblast")
CELL_TYPES <- unique(seurat_object_renamed@meta.data$cell_type)
CELL_TYPES <- sapply(CELL_TYPES, function(x) gsub("_", " ", x))

cell_types <- read.table(file.path(out_dir, "cell_type_annotations.csv"),
                         header = TRUE, stringsAsFactors = FALSE,
                         sep = ",")
# remove unassigned and cancer from further analysis
cell_types <- filter(cell_types, cell_type %in% CELL_TYPES)

# Rename cell types
cell_types$cell_type <- sapply(cell_types$cell_type,
       function(x) gsub("_", " ", x))

# add sample origin
cell_types$origin <- sapply(cell_types$orig.ident, 
                            function(x) samples.tbl[samples.tbl$sample == x, 'origin'])

# add dataset 
cell_types$dataset <- sapply(cell_types$orig.ident, 
                            function(x) samples.tbl[samples.tbl$sample == x, 'dataset'])

# add patient
cell_types$patient <- sapply(cell_types$orig.ident,
                            function(x) samples.tbl[samples.tbl$sample == x, 'patient'])

tbl <- cell_types %>% group_by(dataset, patient, origin, cell_type) %>% summarise(n=n())
Ncells <- group_by(tbl, patient, origin) %>% summarise(NCells = sum(n))
tbl$NCells <- apply(tbl, 1, function(x) Ncells[Ncells$patient == x['patient'] & Ncells$origin == x['origin'], ]$NCells)
tbl <- mutate(tbl, fraction = n / NCells)

write.table(tbl, file.path(out_dir, "table_cell_type_patient_fractions.tsv"),
            quote = FALSE, sep = "\t", row.names = FALSE)
```

```{r}
cols <- c('normal' = '#009E73', 'tumor' = '#D55E00')
ggplot(tbl, aes(x=cell_type, y=fraction, fill=origin)) + 
  geom_bar(stat = "identity", position="dodge") +
  facet_wrap(. ~ patient, ncol = 3) +
  geom_text(aes(label=n), position = position_dodge2(1), vjust=0, hjust=0.5, size=2, angle=0) + 
  labs(y="Fraction of cell type",
       x=NULL,
       title=paste0("Cell type composition for ", sum(tbl$n), " cells" )) +
  theme_bw() + 
  scale_fill_manual(values = cols) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size=10))
ggsave(file.path(out_dir, "bar_cell_type_origin_fractions.png"),
       width = 20, height = 10, units = "cm",
       dpi=300)

# re-order cell type positions
tbl$cell_type <- factor(tbl$cell_type, levels = CELL_TYPES)
ggplot(tbl, aes(x=cell_type, y=fraction, fill=origin)) + 
  geom_bar(stat = "identity", position="dodge") +
  facet_wrap(. ~ patient, ncol = 3) +
  geom_text(aes(label=n), position = position_dodge2(1), hjust=0, size=2, angle=90) + 
  labs(y="Fraction of cell type",
       title=paste0("Cell type composition for ", sum(tbl$n), " cells" )) +
  theme_bw() + 
  scale_fill_manual(values = cols) +
  ylim(c(0,1)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size=10),
        axis.title.x = element_blank())
ggsave(file.path(out_dir, "bar_cell_type_origin_fractions_ordered.png"),
       width = 18, height = 9, units = "cm",
       dpi=600)
```

Plot boxplot of fraction per dataset
```{r}
# rename origin
tbl[tbl$origin == "normal","origin"] <- "control"
cols <- c('control' = '#009E73', 'tumor' = '#D55E00')
tbl$cell_type <- factor(tbl$cell_type, levels = CELL_TYPES)
ggplot(tbl, aes(x=reorder(cell_type, desc(cell_type)), y = fraction, fill = origin)) + 
  geom_boxplot() + 
  facet_grid(origin ~ dataset) +
  labs(y="Fraction of cell type") +
  theme_bw() + 
  coord_flip() +
  scale_fill_manual(values = cols) +
  theme(axis.title.y = element_blank(),
        axis.text = element_text(size = 8),
        axis.title.x = element_text(size = 10, face = "plain", family = "sans"),
        legend.position = "none")
ggsave(file.path(out_dir, "boxplot_cell_type_origin_fractions_ordered.png"),
       width = 8, height = 7, units = "cm")
```

Summarise by dataset
```{r}
tblds <- cell_types %>% group_by(dataset, origin, cell_type) %>% summarise(n=n())
Ncells <- group_by(tblds, dataset, origin) %>% summarise(NCells = sum(n))
tblds$NCells <- apply(tblds, 1, function(x) Ncells[Ncells$dataset == x['dataset'] & Ncells$origin == x['origin'], ]$NCells)
tblds <- mutate(tblds, fraction = n / NCells)
tblds$cell_type <- factor(tblds$cell_type, levels = CELL_TYPES)

ggplot(tblds, aes(x=cell_type, y=fraction, fill=origin)) + 
  geom_bar(stat = "identity", position="dodge") +
  facet_wrap(. ~ dataset, ncol = 3) +
  geom_text(aes(label=n), position = position_dodge2(1), hjust=0, size=2, angle=90) + 
  labs(y="Fraction of cell type",
       title=paste0("Cell type composition for ", sum(tbl$n), " cells" )) +
  theme_bw() + 
  ylim(c(0,0.7)) +
  scale_fill_manual(values = cols) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size=10),
        axis.title.x = element_blank())
ggsave(file.path(out_dir, "bar_cell_type_dataset_fractions_ordered.png"),
       width = 14, height = 8, units = "cm",
       dpi=600)
```

for one dataset
```{r}
CELL_TYPES <- c("T_cell", "B_cell", "mast_cell", "myeloid",
                                                      "alveolar", "epithelial", "endothelial", "fibroblast")

cell_types <- read.table(file.path(out_dir, "cell_type_annotations.csv"),
                         header = TRUE, stringsAsFactors = FALSE,
                         sep = ",")
# remove unassigned and cancer from further analysis
cell_types <- filter(cell_types, cell_type %in% CELL_TYPES)

# add sample origin
cell_types$origin <- sapply(cell_types$orig.ident, 
                            function(x) samples.tbl[samples.tbl$sample == x, 'origin'])

# add dataset 
cell_types$dataset <- sapply(cell_types$orig.ident, 
                            function(x) samples.tbl[samples.tbl$sample == x, 'dataset'])

tbl <- cell_types %>% group_by(origin, cell_type) %>% summarise(n=n())
Ncells <- group_by(tbl, origin) %>% summarise(NCells = sum(n))
tbl$NCells <- apply(tbl, 1, function(x) Ncells[ Ncells$origin == x['origin'], ]$NCells)
tbl <- mutate(tbl, fraction = n / NCells)

write.table(tbl, paste0(out_dir, "table_cell_type_origin_fractions.tsv"),
            quote = FALSE, sep = "\t", row.names = FALSE)
```

```{r}
# re-order cell type positions
tbl$cell_type <- factor(tbl$cell_type, levels = CELL_TYPES)
ggplot(tbl, aes(x=cell_type, y=fraction, fill=origin)) + 
  geom_bar(stat = "identity", position="dodge") +
  geom_text(aes(label=n), position = position_dodge2(1), hjust=0, size=3.5, angle=90) + 
  labs(y="Fraction of cell type",
       title=paste0("Cell type composition for ", sum(tbl$n), " cells" )) +
  theme_bw() + 
  ylim(c(0,0.7)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size=10),
        axis.title.x = element_blank(),
        legend.position = "none")
ggsave(paste0(out_dir, "bar_cell_type_origin_fractions_ordered.png"),
       width = 10, height = 9, units = "cm",
       dpi=600)
```


### plot sum reads
```{r}
ggplot(sum.df, aes(x=cell_type, y = sum_nCount_RNA, fill=cell_type, ymin=0)) + 
  geom_col() +
  facet_wrap(. ~ dataset, scales = "free") + 
  theme_bw() +
  theme(legend.position = "None",
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1, size=10)) +
  labs(y="Sum of UMI counts")
ggsave(file.path(out_dir, "bar_chart_sum_reads.png"),
       width = 16, height = 8, units = "cm")
```

Group by origin
```{r}
cols <- c('normal' = '#009E73', 'tumor' = '#D55E00')
sum.df <- meta.df %>% 
  group_by(dataset, origin, cell_type) %>% 
  summarise(sum_nCount_RNA = sum(nCount_RNA),
                                              n_cells = n())

brks <- seq(1, ceiling(log10(max(sum.df$sum_nCount_RNA))), by = 2)
minbrks <- seq(1, ceiling(log10(max(sum.df$sum_nCount_RNA))), by = 1)

ggplot(sum.df, aes(x=cell_type, y = sum_nCount_RNA, fill=origin, ymin=1)) + 
  geom_bar(stat = "identity", position = "stack") +
  facet_wrap(. ~ dataset, scales = "free", ncol=2) + 
  theme_bw() + mytheme + 
  scale_fill_manual(values = cols) + 
  theme(
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1, size=10),
        legend.position = "none") +
  labs(y="Sum of UMI counts")
ggsave(file.path(out_dir, "bar_chart_sum_reads_per_origin.png"),
       width = 14, height = 8, units = "cm")
```

Group by patient
```{r}
sum.df <- meta.df %>% 
  group_by(patient, origin, cell_type) %>% 
  summarise(sum_nCount_RNA = sum(nCount_RNA),
                                              n_cells = n())

brks <- seq(1, ceiling(log10(max(sum.df$sum_nCount_RNA))), by = 2)
minbrks <- seq(1, ceiling(log10(max(sum.df$sum_nCount_RNA))), by = 1)

ggplot(sum.df, aes(x=cell_type, y = sum_nCount_RNA, fill=origin, ymin=1)) + 
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(. ~ patient, scales = "free", ncol=2) + 
  theme_bw() +
  scale_fill_manual(values = cols) + 
  theme(
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1, size=10),
        legend.position = "none") +
  labs(y="Sum of UMI counts")
ggsave(file.path(out_dir, "bar_chart_sum_reads_per_patient.png"),
       width = 16, height = 18, units = "cm")
```

Save session info
```{r}
writeLines(capture.output(sessionInfo()), file.path(out_dir, "sessionInfo.txt"))
```

